managers/scheduler.py

```diff
@@ -0,0 +1,230 @@
+    mw_id: MaintenanceID
+
+    def __call__(self):
+        self.maintenance_scheduler.start_maintenance(self.mw_id)
+
+
+@dataclass
+class MaintenanceEnd:
+    """
+    Callable used for ending maintenance windows
+    """
+    maintenance_scheduler: 'MaintenanceScheduler'
+    mw_id: MaintenanceID
+
+    def __call__(self):
+        self.maintenance_scheduler.end_maintenance(self.mw_id)
+
+@dataclass
+class MaintenanceScheduler:
+    """Class for scheduling maintenance windows."""
+    deployer: MaintenanceDeployer
+    db_controller: MaintenanceController
+    scheduler: BaseScheduler
+
+    @classmethod
+    def new_scheduler(cls, deployer: MaintenanceDeployer):
+        """
+        Creates a new scheduler from the given MaintenanceDeployer
+        """
+        scheduler = BackgroundScheduler(timezone=pytz.utc)
+        db_controller = MaintenanceController()
+        db_controller.bootstrap_indexes()
+        instance = cls(deployer, db_controller, scheduler)
+        return instance
+
+    def start(self):
+        """
+        Begin running the scheduler.
+        """
+        self.db_controller.prepare_start()
+
+        # Populate the scheduler with all pending tasks
+        windows = self.db_controller.get_windows()
+        for window in windows:
+            if window.status == Status.RUNNING:
+                self.deployer.start_mw(window)
+            self._schedule(window)
+
+        # Start the scheduler
+        self.scheduler.start()
+
+    def shutdown(self):
+        """
+        Stop running the scheduler.
+        """
+        windows = self.db_controller.get_windows()
+
+        # Depopulate the scheduler
+        for window in windows:
+            self._unschedule(window)
+
+        self.scheduler.remove_all_jobs()
+        self.scheduler.shutdown()
+
+    def start_maintenance(self, mw_id: MaintenanceID):
+        """Begins executing the maintenance window
+        """
+        # Get Maintenance from DB and Update
+        window = self.db_controller.start_window(mw_id)
+
+        # Activate Running
+        self.deployer.start_mw(window)
+
+        # Schedule next task
+        self._schedule(window)
+
+    def end_maintenance(self, mw_id: MaintenanceID):
+        """Ends execution of the maintenance window
+        """
+        # Get Maintenance from DB
+        window = self.db_controller.end_window(mw_id)
+
+        # Set to Ending
+        self.deployer.end_mw(window)
+
+    def end_maintenance_early(self, mw_id: MaintenanceID):
+        """Ends execution of the maintenance window early
+        """
+        # Get Maintenance from DB
+        window = self.db_controller.end_window(mw_id)
+
+        # Unschedule tasks
+        self._unschedule(window)
+
+    def add(self, window: MaintenanceWindow, force=False):
+        """Add jobs to start and end a maintenance window."""
+
+        if force is False:
+            overlapping_windows = self.db_controller.check_overlap(window)
```

Comment:
@Ktmi I've just found a corner case on how overlapping is being checked. It's not excluding `finished` maintenance windows, so if a mw is finished earlier with `PATCH /{mw_id}/end`, its `end` datetime would still be considered, which it shouldn't since that `mw` is finished. Could you double check and review the overlapping cases?

Reply:
Should be handled now, check_overlap now checks if the windows is not finished.